core/bcast/recast.go

```diff
@@ -69,21 +72,62 @@ func (r *Recaster) Store(_ context.Context, duty core.Duty,
 	return nil
 }
 
+// StorePregen stores pre-generated aggregate signed duty registrations for rebroadcasting.
+func (r *Recaster) StorePregen(_ context.Context, duty core.Duty,
+	pubkey core.PubKey, aggData core.SignedData,
+) error {
+	if duty.Type != core.DutyBuilderRegistration {
+		return nil
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	tuple, ok := r.pregenTuples[pubkey]
+	if ok && tuple.duty.Slot >= duty.Slot {
+		// Not storing duplicate or older registration.
+		return nil
+	}
+
+	// Clone before storing
+	data, err := aggData.Clone()
+	if err != nil {
+		return err
+	}
+
+	r.pregenTuples[pubkey] = recastTuple{
+		pregenerate: true,
+		duty:        duty,
+		aggData:     data,
+	}
+
+	return nil
+}
+
 // SlotTicked is called when new slots tick.
 func (r *Recaster) SlotTicked(ctx context.Context, slot core.Slot) error {
 	if !slot.FirstInEpoch() {
 		return nil
 	}
 	ctx = log.WithTopic(ctx, "bcast")
 
-	// Copy locked things before doing IO
+	// Copy locked things before doing IO.
 	var (
 		clonedTuples = make(map[core.PubKey]recastTuple)
 		clonedSubs   []func(context.Context, core.Duty, core.PubKey, core.SignedData) error
 	)
 
 	r.mu.Lock()
 	clonedSubs = append(clonedSubs, r.subs...)
+
+	// Populate pre-generated registrations first which can be overridden by VC submitted ones.
```

Comment:
not a fan of all this custom logic for pre-generated registrations. Really do not think recast should even know about this use-case. What about just tracking the latest duty slot we are recasting for each pubkey and then you can query for `slot=0` in grafana to display "pre-generated" registrations.

Reply:
@corverroos do we need to capture broadcast delay for builder registration duty? we calculate broadcast delay from genesis start time:
```
slotStart := genesis.Add(slotDuration * time.Duration(slot))
delay := time.Since(slotStart)
```

If we don't care about the broadcast delay of this duty then I think we are good. Also IMO broadcast delay for this duty isn't that critical.

Reply:
indeed, delay not critical